### DISCUSSION TITLE
Can we make readPreference in @MongoEntity configurable while using mongodb panache ?

### DISCUSSION THREAD

**Discussion by nitinty** [quarkusio/quarkus#43470] [Q&A]
Hi Team,

Seems readPreference is not configurable in @MongoEntity which will require code deployment if we make any changes in readPreference at entity level.

Is there any plan to make it config based so that no code deployment required if we make any change in readPreference ? 

`@MongoEntity(collection = "reactive_user_collection", readPreference = "secondary")`

I can do the required changes if you guys allow me

[a comment by an automated account is not shown]

**Reply by loicmathieu** [reply]
Yes, you can configure it globally inside `quarkus.mongodb.read-preference`.
See https://quarkus.io/guides/mongodb#configuration-reference for all configuration options.

**Reply by nitinty** [reply]
@loicmathieu 
But it would be set for all the mongo entities if we set it globally not for specific one. right ?

**Reply by loicmathieu** [reply]
Yes, we didn't offer any configuration options for MongoDB with Panache, the configuration is at the MongoDB level.

What you can do is to configure a different named MongoDB client (see https://quarkus.io/guides/mongodb#multiple-mongodb-clients), and use `@MongoEntity(client="my-client")`.

**Reply by nitinty** [reply]
But is it good practice to make multiple mongo client just to read from different readPreference ? 

Instead we can make the existing one configurable like below 

<img width="458" alt="Screenshot 2024-09-24 at 9 04 43 PM" src="https://github.com/user-attachments/assets/f75c5126-fbb8-4e42-a673-a6783461e777">

**Reply by loicmathieu** [reply]
This will mix configuration for MongoDB with Panache inside the MongoDB client wich is not only used by MongoDB with Panache so it cannot be done like that.

What will be the expectation of a user of the MongoDB client when he does `mongoClient.getCollection("user_collection")` with this configuration ?

**Reply by nitinty** [reply]
With the above configuration, The user_collection will use a secondary readPreference by default. If you need to switch to primary, you can simply update the configuration, without requiring any code changes at the entity layer. And there is no separate named mongoclient required.


I can add this support in Quarkus framework.

**Reply by loicmathieu** [reply]
I think we don't understand, did you plan to configure that at the MongoDB client level when building the client?

**Reply by nitinty** [reply]
By overriding the behaviour of getCollection method of MongoDatabaseImpl class we can achieve it 

`public <TDocument> MongoCollection<TDocument> getCollection(String collectionName, Class<TDocument> documentClass)`

**Reply by loicmathieu** [reply]
@nitinty sorry to come back to this so late, can you opened a feature request inside the repository so we didn't loose track?
